### PR TITLE
Remove 9.5 support.

### PR DESCRIPTION
**Pull request by anarazel**
This removes 9.5 specific code files, 9.5 conditional code, and the now unneeded CitusNode* wrappers.

This includes parts of #883, namely the duplicated/copied readfuncs/read.c code.

### CODE REVIEW

File: src/test/regress/expected/multi_join_order_additional.out [hunk 1]
```diff
@@ -2,13 +2,6 @@
 -- MULTI_JOIN_ORDER_ADDITIONAL
 --
 ALTER SEQUENCE pg_catalog.pg_dist_shardid_seq RESTART 650000;
--- print whether we're running on 9.5 to make version-specific tests clear
-SELECT substring(version(), '\d+(?:\.\d+)?') = '9.5' AS is_95;
- is_95 
--------
- f
```
Comment (jasonmp85) [inline src/test/regress/expected/multi_join_order_additional.out]: You removed the test file that is _not_ `9.5`?

Comment (jasonmp85) [inline src/test/regress/expected/multi_join_order_additional.out]: Oh. Now I see what happened here. Nevermind!

Comment (anarazel) [inline src/test/regress/expected/multi_join_order_additional.out]: Phew, you shocked me for a second.

File: src/include/distributed/citus_nodefuncs.h [hunk 2]
```diff
@@ -47,19 +35,10 @@ extern void RegisterNodes(void);
...
-#else
-#define READFUNC_ARGS void
-#define READFUNC_RET Node *
-#endif
 
-#if (PG_VERSION_NUM >= 90600)
 #define OUTFUNC_ARGS StringInfo str, const struct ExtensibleNode *raw_node
-#else
-#define OUTFUNC_ARGS StringInfo str, const Node *raw_node
-#endif
```
Comment (anarazel) [inline src/include/distributed/citus_nodefuncs.h]: FWIW, we could remove OUTFUNC_ARGS etc now, but I left that for a separate PR.

Comment (jasonmp85) [inline src/include/distributed/citus_nodefuncs.h]: > left that for a separate PR

Maybe a separate commit? It's a find-and-replace, basically, right?

Comment (anarazel) [inline src/include/distributed/citus_nodefuncs.h]: > Maybe a separate commit?
Can, but not enthusiastic.

> It's a find-and-replace, basically, right?

Yea, pretty much.It makes the lines a good bit longer tho.

File: src/backend/distributed/planner/multi_planner.c [hunk 3]
```diff
@@ -444,7 +444,7 @@ SerializeMultiPlan(MultiPlan *multiPlan)
 	char *serializedMultiPlan = NULL;
 	Const *multiPlanData = NULL;
 
-	serializedMultiPlan = CitusNodeToString(multiPlan);
```
Comment (jasonmp85) [inline src/backend/distributed/planner/multi_planner.c]: Hooray! 💀 to the duplicative `Citus`-prefixed functions!

**Comment by jasonmp85** [comment]
This looks pretty good. Can you do a quick audit of `gitignore` and `gitattributes` to see whether lines can be removed? Are there any documentation references to `9.5`? Code comments?

I'm running some tests with code coverage locally to see whether there is any unexercised code that can be removed (i.e. maybe 9.5-specific code that isn't immediately apparent because it lacks any proximate `PG_VERSION_NUM` check, but isn't just untested code). Otherwise, I think this looks pretty straightforward.

**Comment by anarazel** [comment]
> This looks pretty good. Can you do a quick audit of gitignore and gitattributes to see whether lines can be removed?

Good catch. Fix pushed.

> Are there any documentation references to 9.5?

I'm not sure what doc references there could be in the citus repo?

> Code comments?

There's one remaining 9.5 reference, but I think that can stay:
```
	 * Look through contain_mutable_functions_walker or future PG's equivalent for new
	 * node types before bumping this version number to fix compilation; e.g. for any
	 * PostgreSQL after 9.5, see check_functions_in_node.
```

**Comment by jasonmp85** [comment]
> I'm not sure what doc references there could be in the citus repo?

`CONTRIBUTING.md`, apparently…

> There's one remaining 9.5 reference

The language around _or future PG's equivalent_ and _PostgreSQL after 9.5_ is extraneous now and should just be removed; can you rewrite it to just directly reference the function implementations we need to check in PostgreSQL 9.6 and 10? The rest is just noise with 9.5 gone.

**Comment by anarazel** [comment]
>>  I'm not sure what doc references there could be in the citus repo?
> CONTRIBUTING.md, apparently…

Ah, right - I'd only searched src/{include,backend} for references, because there's  "9.5" strings in some of the test output, ignoring files would have been better.  Also found one more reference in pg_regress_multi.pl.  Pushed fixes for these.

> The language around or future PG's equivalent and PostgreSQL after 9.5 is extraneous now and should just be removed; can you rewrite it to just directly reference the function implementations we need to check in PostgreSQL 9.6 and 10? The rest is just noise with 9.5 gone.

Seems like pointless diff noise to me.

**Comment by jasonmp85** [comment]
OK, so I've audited other unused code and including here for completeness…

# Things Enabled by Removing 9.5

  * Kill off our `RangeTblFunction` hack in favor of always using custom nodes (tracked in #703)
  * Remove some `ruleutils` modifications after that
  * Remove whitelisted distributed aggregates and use parallel aggregate functions (tracked in #120)

# Dead Code

Not sure if always dead or now dead because of 9.5 changes. Checked rebalancer/enterprise for calls and found none:

 * `LowestShardIntervalById`
 * `UnlockShardResource`
 * `LockMetadataSnapshot`
 * `OpExpressionContainsColumn`

# Deprecated Paths

The workerlist file parsing code and many functions within `worker_node_manager` seem overdue for killing off.

**Comment by jasonmp85** [comment]
> Seems like pointless diff noise to me.

Having read that comment a couple of times and always having to make it to the very end to actually get the information I'm after, I disagree (people are going to read the comment more often than look at `git blame`), but not a big deal; would just be nice to have precisely the info we need for the versions we support, and have it frontloaded in that sentence.

No other feedback after my `lcov` runs.

**Comment by anarazel** [comment]
> Not sure if always dead or now dead because of 9.5 changes. Checked rebalancer/enterprise for calls and found none:

Don't think any of these are because of the 9.5 removal.

> LowestShardIntervalById

1585c0232277e0a02952105116f398df037993cb

Seems not unlikely to gain a new caller again.

> UnlockShardResource

Don't think we ever had a caller, that's more completeness I think. I'd keep it.

> LockMetadataSnapshot

92c756700831392aa901ec67636548a72e58b490

Hm. Unclear. @marcocitus ?

> OpExpressionContainsColumn

d399f395f7c409fc43a1affcd37a127557e1bb5c

I think this one should be removed.

**Comment by anarazel** [comment]
Thanks for the review!

**Pull request merged**